File: go-service/internal/interfaces/metrics_handler.go
package interfaces

import (
	"fmt"
	"net/http"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

// RequestTrace represents a single HTTP request trace captured by the middleware.
type RequestTrace struct {
	Method     string  `json:"method"`
	Path       string  `json:"path"`
	Status     int     `json:"status"`
	DurationMs float64 `json:"duration_ms"`
	Timestamp  string  `json:"timestamp"`
}

// MetricsHandler tracks request and error counters using thread-safe atomic values,
// and maintains an append-only slice of request traces.
type MetricsHandler struct {
	requestsTotal atomic.Int64
	errorsTotal   atomic.Int64
	traces        []RequestTrace
	mu            sync.Mutex
}

// NewMetricsHandler creates a new MetricsHandler with zero-initialized counters
// and an empty append-only slice for traces.
func NewMetricsHandler() *MetricsHandler {
	return &MetricsHandler{
		traces: make([]RequestTrace, 0),
	}
}

// PushTrace appends a trace to the append-only slice. This method is thread-safe.
func (m *MetricsHandler) PushTrace(trace RequestTrace) {
	m.mu.Lock()
	m.traces = append(m.traces, trace)
	m.mu.Unlock()
}

// GetTraces returns up to `limit` traces from the append-only slice, newest-first.
// If fewer traces exist than `limit`, all available traces are returned.
// This method is thread-safe.
func (m *MetricsHandler) GetTraces(limit int) []RequestTrace {
	m.mu.Lock()
	defer m.mu.Unlock()

	n := len(m.traces)
	if n == 0 {
		return []RequestTrace{}
	}
	if limit > n {
		limit = n
	}
	if limit < 0 {
		limit = 0
	}

	result := make([]RequestTrace, limit)
	for i := 0; i < limit; i++ {
		result[i] = m.traces[n-1-i] // newest-first
	}
	return result
}

// parseLimit parses the "limit" query parameter, returning the default value if
// the parameter is missing or not a valid positive integer.
func parseLimit(r *http.Request, defaultLimit int) int {
	limitStr := r.URL.Query().Get("limit")
	if limitStr == "" {
		return defaultLimit
	}
	var limit int
	if _, err := fmt.Sscanf(limitStr, "%d", &limit); err != nil {
		return defaultLimit
	}
	if limit < 1 {
		return 1
	}
	if limit > 10000 {
		return 10000
	}
	return limit
}

// HandleRequests handles GET /metrics/requests requests, returning JSON with
// requests_total, errors_total, and recent traces (newest-first, clamped by limit).
func (m *MetricsHandler) HandleRequests(w http.ResponseWriter, r *http.Request) {
	limit := parseLimit(r, 50)
	traces := m.GetTraces(limit)

	writeJSON(w, http.StatusOK, map[string]interface{}{
		"requests_total": m.requestsTotal.Load(),
		"errors_total":   m.errorsTotal.Load(),
		"recent":         traces,
	})
}

// ServeHTTP handles GET /metrics requests and returns JSON with counter values.
func (m *MetricsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, map[string]int64{
		"requests_total": m.requestsTotal.Load(),
		"errors_total":   m.errorsTotal.Load(),
	})
}

// IncrementRequests increments the requests counter by 1.
func (m *MetricsHandler) IncrementRequests() {
	m.requestsTotal.Add(1)
}

// IncrementErrors increments the errors counter by 1.
func (m *MetricsHandler) IncrementErrors() {
	m.errorsTotal.Add(1)
}

// MetricsMiddleware returns an HTTP middleware that counts every request through
// the requests counter, counts errors (status >= 400) through the errors counter,
// and captures a RequestTrace for each request into the append-only slice.
// The returned factory matches the existing middleware pattern in this package:
// it takes an http.Handler and returns the wrapped handler.
func MetricsMiddleware(m *MetricsHandler) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Skip health and metrics endpoints — they're infrastructure noise, not business traffic
			if strings.HasPrefix(r.URL.Path, "/health") || strings.HasPrefix(r.URL.Path, "/metrics") {
				next.ServeHTTP(w, r)
				return
			}

			start := time.Now()

			m.IncrementRequests()

			rw := &responseWriter{ResponseWriter: w, statusCode: http.StatusOK}
			next.ServeHTTP(rw, r)

			if rw.statusCode >= 400 {
				m.IncrementErrors()
			}

			m.PushTrace(RequestTrace{
				Method:     r.Method,
				Path:       r.URL.Path,
				Status:     rw.statusCode,
				DurationMs: time.Since(start).Seconds() * 1000,
				Timestamp:  time.Now().UTC().Format(time.RFC3339),
			})
		})
	}
}

File: node-service/src/interfaces/metrics-handler.js
/**
 * In-memory metrics handler with request and error counters.
 *
 * Thread-safe not needed — Node.js is single-threaded for JS execution.
 */
export class MetricsHandler {
  constructor() {
    /** @type {number} */
    this.requests = 0;
    /** @type {number} */
    this.errors = 0;
    /** Ring buffer capacity. */
    this.cap = 200;
    /** @type {Array<{method: string, path: string, status: number, duration_ms: number, timestamp: string}>} */
    this.traces = new Array(this.cap);
    /** @type {number} */
    this.writeIdx = 0;
    /** @type {number} */
    this.count = 0;
  }

  /**
   * Push a request trace into the ring buffer.
   * Overwrites oldest entry when at capacity.
   *
   * @param {{method: string, path: string, status: number, durationMs: number, timestamp: string}} trace
   */
  pushTrace({ method, path, status, durationMs, timestamp }) {
    this.traces[this.writeIdx] = { method, path, status, duration_ms: durationMs, timestamp };
    this.writeIdx = (this.writeIdx + 1) % this.cap;
    this.count++;
  }

  /**
   * Retrieve up to `limit` traces, newest-first.
   *
   * @param {number} limit
   * @returns {Array<{method: string, path: string, status: number, duration_ms: number, timestamp: string}>}
   */
  getTraces(limit) {
    const stored = Math.min(this.count, this.cap);
    if (stored === 0) return [];
    const clamped = Math.max(1, Math.min(limit, this.cap));
    const resultSize = Math.min(clamped, stored);
    const result = new Array(resultSize);
    for (let i = 0; i < resultSize; i++) {
      const idx = (this.writeIdx - 1 - i + this.cap) % this.cap;
      result[i] = { ...this.traces[idx] };
    }
    return result;
  }

  /** Increment the total request counter. */
  incrementRequest() {
    this.requests++;
  }

  /** Increment the total error counter. */
  incrementError() {
    this.errors++;
  }

  /**
   * Parse limit from query string, clamped to [1, cap].
   *
   * @param {string} urlStr
   * @returns {number}
   */
  _parseLimit(urlStr) {
    const params = new URL(urlStr, 'http://localhost').searchParams;
    const raw = params.get('limit');
    const limit = raw !== null ? parseInt(raw, 10) : 50;
    if (Number.isNaN(limit) || limit < 1) return 1;
    return Math.min(limit, this.cap);
  }

  /**
   * Handle GET /metrics/requests — returns JSON with counters and recent traces.
   *
   * @param {import('node:http').IncomingMessage} req
   * @param {import('node:http').ServerResponse} res
   */
  handleRequests(req, res) {
    const limit = this._parseLimit(req.url);
    const traces = this.getTraces(limit);

    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(
      JSON.stringify({
        requests_total: this.requests,
        errors_total: this.errors,
        recent: traces,
      }) + '\n',
    );
  }

  /**
   * Handle GET /metrics — returns JSON with accumulated counters.
   *
   * @param {import('node:http').IncomingMessage} _req
   * @param {import('node:http').ServerResponse} res
   */
  handleMetrics(_req, res) {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(
      JSON.stringify({
        requests_total: this.requests,
        errors_total: this.errors,
      }) + '\n',
    );
  }
}
